lib/OpenQA/Worker/Job.pm

```diff
@@ -0,0 +1,1263 @@
+            next if $retry_counter;
+
+            # Just return if all upload retries have failed
+            # this will cause the next group of uploads to be triggered
+            my $msg = "All $retry_limit upload attempts have failed for $filename";
+            log_error($msg, channels => ['autoinst', 'worker'], default => 1);
+            return 0;
+        }
+        last;
+    }
+
+    return 0 if _log_upload_error($filename, $res);
+    return 1;
+}
+
+sub _read_json_file {
+    my ($self, $name) = @_;
+
+    my $pooldir = $self->worker->pool_directory;
+    my $fn      = "$pooldir/testresults/$name";
+    local $/;
+    my $fh;
+    if (!open($fh, '<', $fn)) {
+        log_warning(
+"Can't open $fn for result upload - likely isotovideo could not be started or failed early. Error message: $!"
+        );
+        return undef;
+    }
+    my $json = {};
+    eval { $json = decode_json(<$fh>); };
+    log_warning("os-autoinst didn't write proper $fn") if $@;
+    close($fh);
+    return $json;
+}
+
+sub _read_result_file {
+    my ($self, $upload_up_to, $extra_test_order) = @_;
+
+    my %ret;
+
+    # upload all results not yet uploaded - and stop at $upload_up_to
+    # if $upload_up_to is empty string, then upload everything
+    my $test_order = $self->test_order;
+    while (my $remaining_test_count = scalar(@$test_order)) {
+        my $test   = $test_order->[0]->{name};
+        my $result = $self->_read_module_result($test);
+
+        my $current_test_module         = $self->current_test_module;
+        my $is_last_test_to_be_uploaded = $remaining_test_count eq 1 || $test eq $upload_up_to;
+        my $test_not_running            = !$current_test_module || $test ne $current_test_module;
+        my $test_is_completed           = !$is_last_test_to_be_uploaded || $test_not_running;
+        if ($test_is_completed) {
+            # remove completed tests from @$test_order so we don't upload those results twice
+            shift(@$test_order);
+        }
+
+        last unless ($result);
+        $ret{$test} = $result;
+
+        if ($result->{extra_test_results}) {
+            for my $extra_test (@{$result->{extra_test_results}}) {
+                my $extra_result = read_module_result($extra_test->{name});
+                next unless $extra_result;
+                $ret{$extra_test->{name}} = $extra_result;
+            }
+            push @{$extra_test_order}, @{$result->{extra_test_results}};
+        }
+
+        last if $is_last_test_to_be_uploaded;
+    }
+    return \%ret;
+}
+
+sub _read_module_result {
+    my ($self, $test) = @_;
+
+    my $result = $self->_read_json_file("result-$test.json");
+    return unless $result;
+
+    my $images_to_send = $self->images_to_send;
+    my $files_to_send  = $self->files_to_send;
+    for my $d (@{$result->{details}}) {
+        if ($d->{json}) {
+            $d->{json} = $d->{json};
+        }
+        for my $n (@{$d->{needles}}) {
+            $n->{json} = $n->{json};
```

Comment:
uh, wut? Are you really overwriting the value with itself?

Reply:
Same 3 lines above.